tools/cfg/bin_to_dot.cpp

```diff
@@ -0,0 +1,165 @@
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include "bin_to_dot.h"
+
+#include <cassert>
+#include <iostream>
+#include <vector>
+
+#include "assembly_grammar.h"
+
+namespace {
+
+// A DotConverter can be used to dump the GraphViz "dot" graph for
+// a SPIR-V module.
+class DotConverter {
+ public:
+  DotConverter(std::iostream* out) : out_(*out) {}
+
+  // Emits the graph preamble.
+  void Begin() const { out_ << "digraph {\n"; }
+  // Emits the graph postamble.
+  void End() const { out_ << "}\n"; }
+
+  // Emits the Dot commands for the given instruction.
+  spv_result_t HandleInstruction(const spv_parsed_instruction_t& inst);
+
+ private:
+  // Ends processing for the current block, emitting its dot code.
+  void FlushBlock(const std::vector<uint32_t>& successors);
+
+  // The ID of the current basic block, or 0 if outside of a block.
+  uint32_t current_block_id_ = 0;
+
+  // Have we completed processing for the entry block to this fuction?
+  bool seen_function_entry_block_ = false;
+
+  // The Id of the merge block for this block if it exists, or 0 otherwise.
+  uint32_t merge_ = 0;
+  // The Id of the continue block for this block if it exists, or 0 otherwise.
+  uint32_t continue_target_ = 0;
+
+  // The output stream.
+  std::ostream& out_;
+};
+
+spv_result_t DotConverter::HandleInstruction(
+    const spv_parsed_instruction_t& inst) {
+  switch(inst.opcode) {
+    case SpvOpFunction:
+      seen_function_entry_block_ = false;
+      break;
+
+    case SpvOpLabel:
+      current_block_id_ = inst.result_id;
+      break;
+
+    case SpvOpBranch:
+      FlushBlock({inst.words[1]});
+      break;
+    case SpvOpBranchConditional:
+      FlushBlock({inst.words[2], inst.words[3]});
+      break;
+    case SpvOpSwitch: {
+      std::vector<uint32_t> successors{inst.words[2]};
+      for (size_t i = 3; i < inst.num_operands; i += 2) {
+        successors.push_back(inst.words[inst.operands[i].offset]);
+      }
+      FlushBlock(successors);
+    } break;
+
+    case SpvOpKill:
+    case SpvOpReturn:
+    case SpvOpUnreachable:
+    case SpvOpReturnValue:
+      FlushBlock({});
+      break;
+
+    case SpvOpLoopMerge:
+      merge_ = inst.words[1];
+      continue_target_ = inst.words[2];
+      break;
+    case SpvOpSelectionMerge:
+      merge_ = inst.words[1];
+      break;
+    default:
+      break;
+  }
+  return SPV_SUCCESS;
+}
+
```

Comment:
It would be nice to add a comment to explain those `shape`s and `style`s in the output.